stl/inc/algorithm

```diff
@@ -961,6 +941,340 @@ namespace ranges {
 } // namespace ranges
 #endif // __cpp_lib_concepts
 
+// FUNCTION TEMPLATE is_permutation
+template <class _FwdIt1, class _FwdIt2, class _Pr>
+_NODISCARD _CONSTEXPR20 bool _Is_permutation_unchecked(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2, _Pr _Pred) {
+    // test if [_First1, _Last1) == permuted [_First2, ...), using _Pred
+    for (; _First1 != _Last1; ++_First1, (void) ++_First2) { // trim matching prefix
+        if (!_Pred(*_First1, *_First2)) {
+            // found first inequality, check match counts in suffix
+            //
+            // narrowing _Iter_diff_t<_FwdIt1> to _Iter_diff_t<_FwdIt2> is OK because if the second range is shorter
+            // than the first, the user already triggered UB
+            auto _Last2 = _STD next(_First2, static_cast<_Iter_diff_t<_FwdIt2>>(_STD distance(_First1, _Last1)));
+            return _Check_match_counts(_First1, _Last1, _First2, _Last2, _Pred);
+        }
+    }
+
+    return true;
+}
+
+template <class _FwdIt1, class _FwdIt2, class _Pr>
+_NODISCARD _CONSTEXPR20 bool is_permutation(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2, _Pr _Pred) {
+    // test if [_First1, _Last1) == permuted [_First2, ...), using _Pred
+    _Adl_verify_range(_First1, _Last1);
+    const auto _UFirst1 = _Get_unwrapped(_First1);
+    const auto _ULast1  = _Get_unwrapped(_Last1);
+    const auto _UFirst2 = _Get_unwrapped_n(_First2, _Idl_distance<_FwdIt1>(_UFirst1, _ULast1));
+    return _Is_permutation_unchecked(_UFirst1, _ULast1, _UFirst2, _Pass_fn(_Pred));
+}
+
+template <class _FwdIt1, class _FwdIt2>
+_NODISCARD _CONSTEXPR20 bool is_permutation(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2) {
+    // test if [_First1, _Last1) == permuted [_First2, ...)
+    return _STD is_permutation(_First1, _Last1, _First2, equal_to<>());
+}
+
+template <class _FwdIt1, class _FwdIt2, class _Pr>
+_CONSTEXPR20 bool _Is_permutation_unchecked(_FwdIt1 _First1, _FwdIt1 _Last1, _FwdIt2 _First2, _FwdIt2 _Last2, _Pr _Pred,
+    forward_iterator_tag, forward_iterator_tag) {

```

Comment:
Would it make sense to define something like `common_iterator_category`

Given that it only appears rarely I guess not too much but meh

Reply:
Technically `common_type` will work, at the cost of some class template instantiations - it just doesn't seem worth the throughput cost to simplify these "two-tag" dispatch functions.